### PR TITLE
Update segmentation algorithm

### PR DESCRIPTION
**Pull request by LDZuckerman**
## PR Description

This pull request would update the segmentation algorithm in `sunkit-image/granule.py` that I submitted around six months ago. After working with a local version of this code for some time, I realized that I had made substantive changes which the sunkit-image version might benefit from. These changes include:

1. Speeding up computational time for large images by computing the initial thresholding based on a random subset of pixels.
2. Increasing accuracy on images with spatially varying background flux levels by applying a local histogram equalization  before computing the initial thresholding. 
3. Preventing errors finding "dim centers" in images that have all-granule edges by adding a "padding" of zero pixels around the edges of the image first.
4. Improving accuracy in brightpoint selection by updating the selection method. The new method first creates a mask of "candidate" brightpoints based on a minimum flux value determined from the equalized image. Then it checks not only which of these regions are small, but also which contain high gradient values in the original (non-equalized) image. 
5. Correcting (flipping) the assignment of the values 2 and 3 to brightpoints and dim centers. Somehow, in the current version, the values actually assigned were opposite what was described in the docstring for `segment()`, and to how the colors were assigned in the final output map. I am also open to creating consistency in the other way, by changing the docstring and the output map colors. 

I considered opening an issue first to discuss these changes, but thought it might be better to share code earlier rather than later. 
Please let me know if these updates seem resonable, and whether (at least in some capacity), you would like to add them.

### CODE REVIEW

**Comment by nabobalis** [comment]
Hi @LDZuckerman, 

Thanks again for coming back to this and improving it. 

The changes look good to me, just need to deal with the failing CI and the style check.

A changelog entry detailing the improvements as well would be great.

File: sunkit_image/granule.py [hunk 1]
```diff
@@ -90,6 +94,8 @@ def _get_threshold(data, method):
     """
     if not isinstance(data, np.ndarray):
         raise ValueError("Input data must be an instance of a np.ndarray")
+    if len(data.flatten()) > 500**2:
+        data = np.random.choice(data.flatten(), (500, 500)) # Computing threshold based on random sample works well and saves significant computatonal time
```
Comment (nabobalis) [inline sunkit_image/granule.py:98]: I think a logging.info message to alert the user to this would be good.

File: sunkit_image/granule.py [hunk 2]
```diff
@@ -128,35 +134,43 @@ def _trim_intergranules(segmented_image, mark=False):
...
                 if not mark:
                     segmented_image_fixed[labeled_seg == value] = 1
                 elif mark:
-                    segmented_image_fixed[labeled_seg == value] = 2
+                    segmented_image_fixed[labeled_seg == value] = 3
     return segmented_image_fixed
 
 
-def _mark_brightpoint(segmented_image, data, resolution, bp_min_flux=None):
+def _mark_brightpoint(segmented_image, data, HE_data, resolution, bp_min_flux=None):
```
Comment (nabobalis) [inline sunkit_image/granule.py:162]:
```suggestion
def _mark_brightpoint(segmented_image, data, he_data, resolution, bp_min_flux=None):
```
and below.

File: sunkit_image/granule.py [hunk 3]
```diff
@@ -166,44 +180,47 @@ def _mark_brightpoint(segmented_image, data, resolution, bp_min_flux=None):
...
     )
-    bp_pix_upper_limit = bp_size_limit / (resolution**2)
+    bp_pix_upper_limit = (bp_size_limit / resolution)**2 # Max area in pixels
     bp_pix_lower_limit = 4  # Very small bright regions are likely artifacts
-    # General flux limit determined by visual inspection.
+    # General flux limit determined by visual inspection (set using equalized map)
     if bp_min_flux is None:
-        stand_devs = 0.5
-        bp_brightness_limit = np.mean(data) + stand_devs * np.std(data)
+        stand_devs = 1.25
```
Comment (nabobalis) [inline sunkit_image/granule.py:197]: What was the reason to increase this?

File: sunkit_image/granule.py (hunk 2 above)
Comment (LDZuckerman) [reply sunkit_image/granule.py:162]: Why the lower case "he"? Do you prefer it that way whenever I use that abbreviation or just here?

File: sunkit_image/granule.py (hunk 3 above)
Comment (LDZuckerman) [reply sunkit_image/granule.py:197]: The minimum flux is now based on the histogram equalized image, and I found that 1.25*SD was what worked most consistantly on those images.

**Comment by LDZuckerman** [comment]
Hmmm... I am having a little trouble determining which tests have failed. What does "`Run python -m tox-e py310`" refer to?

File: sunkit_image/granule.py (hunk 3 above)
Comment (nabobalis) [reply sunkit_image/granule.py:198]: Can you leave a comment as to that here jus for future reference please?

**Comment by nabobalis** [comment]
> Hmmm... I am having a little trouble determining which tests have failed. What does "`Run python -m tox-e py310`" refer to?

That refers to the command that is run, if you open that section, you should see the full test trace.

If you see https://github.com/sunpy/sunkit-image/actions/runs/5746471664/job/15575996325?pr=154#step:10:284 for example

File: sunkit_image/granule.py (hunk 2 above)
Comment (nabobalis) [reply sunkit_image/granule.py:163]: I personally don't like variables to have mixed cases is all.

**Comment by nabobalis** [comment]
> Sorry - I seem to be getting some user warnings regarding the data type in `skimage.filters.rank.equalize()` that I am not getting when I run this locally. I will figure it out.

Most likely that the version of scikit image being run on the CI is newer than your local version. IF you upgrade it, hopefully it will raise the same warning.

File: sunkit_image/tests/conftest.py [hunk 4]
```diff
@@ -95,6 +96,13 @@ def pytest_runtest_setup(item):
 def granule_map():
     return sunpy.map.Map(get_pkg_data_filename("dkist_photosphere.fits", package="sunkit_image.data.test"))
 
+@pytest.fixture()
+def granule_map_he():
+    granule_map = sunpy.map.Map(get_pkg_data_filename("dkist_photosphere.fits", package="sunkit_image.data.test"))
+    map = granule_map.data
+    map_norm = ((map - np.nanmin(map))/(np.nanmax(map) - np.nanmin(map))) # min-max normalization to [0, 1]
```
Comment (nabobalis) [inline sunkit_image/tests/conftest.py:103]: Is there no helper in scikit image to do the normalization to [0, 1]?

File: changelog/154.feature.rst [hunk 5]
```diff
@@ -0,0 +1,7 @@
+Modifcations to the `sunkit_image.granule` module. 
+1. Increase in speed for large images achieved by computing the initial thresholding on a random subset of pixels.
+2. Increase in accuracy on images with spatially varying background flux levels achieved by applying a local histogram 
+equalization before computing the initial thresholding.
+3. Prevention of errors in finding "dim centers" in images that have all-granule edges achieved by adding a "padding" 
+of zero pixels around the edges.
+5. Correction of the assignment of the values 2 and 3 to brightpoints and dim centers.
```
Comment (nabobalis) [inline changelog/154.feature.rst:7]:
```suggestion
Modifcations to the `sunkit_image.granule` module. 

1. Increase in speed for large images achieved by computing the initial thresholding on a random subset of pixels.
2. Increase accuracy on images with spatially varying background flux levels achieved by applying a local histogram equalization before computing the initial thresholding.
3. Prevention of errors in finding "dim centers" in images that have all-granule edges achieved by adding a "padding" of zero pixels around the edges.
5. Correction of the assignment of the values 2 and 3 to brightpoints and dim centers. 
```

File: sunkit_image/granule.py [hunk 6]
```diff
@@ -52,15 +52,19 @@ def segment(smap, *, skimage_method="li", mark_dim_centers=False, bp_min_flux=No
         resolution = smap.scale[0].value
     else:
         raise ValueError("Currently only maps with square pixels are supported.")
-    median_filtered = scipy.ndimage.median_filter(smap.data, size=3)
+    # Obtain local histogram equalization of map.
+    map = smap.data
```
Comment (nabobalis) [inline sunkit_image/granule.py:56]: Do we need to do assign this variable? IF so, can it be renamed to not, "map".

File: sunkit_image/granule.py [hunk 7]
```diff
@@ -90,6 +94,9 @@ def _get_threshold(data, method):
     """
     if not isinstance(data, np.ndarray):
         raise ValueError("Input data must be an instance of a np.ndarray")
+    if len(data.flatten()) > 500**2:
+        logging.info("Image is large, so threshold computation will be based on a random 500x500 sample of pixels")
```
Comment (nabobalis) [inline sunkit_image/granule.py:98]:
```suggestion
        logging.info("Input image is large (> 500**2), so threshold computation will be based on a random 500x500 sample of pixels")
```

File: sunkit_image/tests/conftest.py (hunk 4 above)
Comment (LDZuckerman) [reply sunkit_image/tests/conftest.py:103]: Hmm... I am not seeing one. I could use an sklearn or scipy package instead.

**Comment by LDZuckerman** [comment]
The test_mark_brightpoint is failing, and I think that is becasue the example image we were using atually _shouldn't_ contain any brightpoints. I am going to update it to a better image.

**Comment by LDZuckerman** [comment]
Looks like the only tests now failing are not ones I wrote, e.g. they are in `sunkit-image/sunkit-image/tests/test_enhance.py` and `sunkit-image/utils/tests/test_noise.py`. Do I need to worry about those?

**Comment by nabobalis** [comment]
If they aren't granule tests, ignore it.

**Comment by LDZuckerman** [comment]
> If they aren't granule tests, ignore it.

Great. Is there anything else that looks problematic or that I should change?

**Comment by nabobalis** [comment]
> > If they aren't granule tests, ignore it.
> 
> Great. Is there anything else that looks problematic or that I should change? 

I'll have to review it again later this week but probably not! 

Thanks again for working on this.

**Review by nabobalis — APPROVED** [review]

**Review by nabobalis — APPROVED** [review]

**Comment by nabobalis** [comment]
Test failures are to be fixed in another PR, these are not due to the changes here.

**Comment by nabobalis** [comment]
Thank you again @LDZuckerman for working on this!

**Pull request merged**